adapters/ogury/ogury.go

```diff
@@ -0,0 +1,209 @@
+				Message: "Bidder extension not provided or can't be unmarshalled",
+			})
+		}
+		// find Ogury bidder params
+		if bidder, ok := impExt[openrtb_ext.PrebidExtBidderKey]; ok {
+			if err := json.Unmarshal(bidder, &impExtBidderHoist); err != nil {
+				return nil, append(errors, &errortypes.BadInput{
+					Message: "Ogury bidder extension not provided or can't be unmarshalled",
+				})
+			}
+		}
+
+		newImpExt := make(map[string]any, len(impExt)-1+len(impExtBidderHoist))
+
+		// copy every imp.ext field to the new "ext" object except for imp.ext.bidder
+		for key, value := range impExt {
+			if key != openrtb_ext.PrebidExtBidderKey {
+				newImpExt[key] = value
+			}
+		}
+		// extract Ogury params from imp.ext.bidder to imp.ext
+		for key, value := range impExtBidderHoist {
+			newImpExt[key] = value
+		}
+
+		ext, err := json.Marshal(newImpExt)
+		if err != nil {
+			return nil, append(errors, &errortypes.BadInput{
+				Message: "Error while marshaling Imp.Ext bidder exension",
+			})
+		}
+		request.Imp[i].Ext = ext
+
+		// save adUnitCode
+		request.Imp[i].TagID = imp.ID
+	}
+
+	// currency conversion
+	for i, imp := range request.Imp {
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if imp.BidFloor > 0 && imp.BidFloorCur != "" && strings.ToUpper(imp.BidFloorCur) != "USD" {
+
+			// Convert to US dollars
+			convertedValue, err := requestInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, "USD")
+			if err != nil {
+				return nil, []error{err}
+			}
+
+			// Update after conversion. All imp elements inside request.Imp are shallow copies
+			// therefore, their non-pointer values are not shared memory and are safe to modify.
+			request.Imp[i].BidFloorCur = "USD"
+			request.Imp[i].BidFloor = convertedValue
+		}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: headers,
+		ImpIDs:  openrtb_ext.GetImpIDs(request.Imp),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+
+}
+
+func filterValidImps(request *openrtb2.BidRequest) (validImps []openrtb2.Imp) {
+	for _, imp := range request.Imp {
+		var impExt adapters.ExtImpBidder
+		var impExtOgury openrtb_ext.ImpExtOgury
+
+		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+			continue
+		}
+		if err := json.Unmarshal(impExt.Bidder, &impExtOgury); err != nil {
+			continue
+		}
+		if impExtOgury.AssetKey != "" && impExtOgury.AdUnitID != "" {
+			validImps = append(validImps, imp)
+		}
+	}
+
+	// if we have imp with assetKey/adUnitId then we want to serve them
+	if len(validImps) > 0 {
+		return validImps
+	}
+
+	// no assetKey/adUnitId imps then we serve everything if publisher.ID exists
+	if request.Site != nil && request.Site.Publisher.ID != "" {
+		return request.Imp
+	}
+
+	// else no valid imp
+	return nil
+}
+
+func setHeaders(request *openrtb2.BidRequest) http.Header {
```

Comment:
Nitpick: This does not set headers, but rather creates a collection of headers. Consider renaming to something like `buildHeaders`.